### PR TITLE
Add support for migrating CHILD_SUBREAPER prctl

### PR DESCRIPTION
**Pull request by mclapinski**
1. Checkpoint it via parasite.
2. Restore it before forking.
3. Add ZDTM test.

### CODE REVIEW

**Comment by rst0git** [comment]
Hi @mclapinski, would it be possible to add a ZDTM test for this feature to make sure it will not break with some changes in the future?

**Comment by mclapinski** [comment]
Hi @rst0git, done.

**Comment by 0x7f454c46** [comment]
> 2. Restore it before forking.

Not sure about that.. I believe we set SIGCHILD handler in criu to catch bugs or unexpected deaths of children. Look at `sigchld_handler()` in `criu/cr-restore.c`. The main part (except reporting) is `futex_abort_and_wake()`, which will stop restore and prevent criu from hanging in such cases.

So, we either need to add SIGCHILD handling to the sub-reaper being restored (which is problematic, as it might be a process, that wasn't C/R'ed), or restore the subreaper much later - even so it will add a prctl-per-thread.

Cc'ing @Snorch as he has eaten a bunch of children, subreapers and all those fantastic animals at some point.

**Comment by mclapinski** [comment]
@0x7f454c46 
Restoring child subreaper attribute after forking will work only on kernels containing this [fix](https://github.com/torvalds/linux/commit/749860ce242798fb090557a5a7868dee40af9268) so starting from v4.11. For older kernels restoring this attribute after forking will have no effect on children this process already has.

**Comment by mclapinski** [comment]
@0x7f454c46 
Sorry but I'm new here. I don't understand what you mean by `sub-reaper being restored (which is problematic, as it might be a process, that wasn't C/R'ed)`. To me this PR is about C/Ring the subreaper process and its children, so I don't really understand how it could not have been C/R'ed.

**Comment by 0x7f454c46** [comment]
@mclapinski one can `criu dump -t <pid>` which has some foreign subreaper already

**Comment by mclapinski** [comment]
> @mclapinski one can `criu dump -t <pid>` which has some foreign subreaper already

Sure, but you said `sub-reaper being restored`, so I got lost. Is the subreaper process being C/R'ed or is it foreign and not being C/R'ed?

**Comment by mclapinski** [comment]
I see it like this:
1. We are C/Ring a process that is the subreaper, so running `PR_GET_CHILD_SUBREAPER` will return `1`. In that case we have to restore that property before forking due to the bug in kernels older than 4.11
2. We are C/Ring a process that isn't the subreaper but one of it ancestors is. In that case we don't care, because child subreaper property is assigned to that ancestor not to us, so there's nothing we should do.

**Comment by 0x7f454c46** [comment]
Ok, let me try again: I like (1) restoring at the forking stage, it looks neat and simple-designed.
But for (2): how will criu acknowledge that a child has unexpectedly died (killed by user/OOM/segfaulted)? As per my poor understanding, SIGCHLD will be delivered to subreaper and criu will just hang on futex waiting for nr_in_progress to cease.

**Comment by mclapinski** [comment]
But what does that have to do with that PR? I'm changing the behaviour of C/R of process that is a child subreaper, not of a process that has a child subreaper among its ancestors. Behaviour in (2) doesn't change with this PR. In (2) parasite runs `PR_GET_CHILD_SUBREAPER`, it returns `0` (meaning we are not a subreaper) and we set it to `0` during restore.

**Comment by 0x7f454c46** [comment]
@mclapinski Oh, I see. Sorry, by some reason I got confused by what `prctl()` interface does.
Thanks for describing. Now I understand implementation a bit better and it looks quite good to me.

Though, I got another question: what will happen if during restore we get unexpected SIGCHLD from side-activity? (Like a task dies after being reparented to criu). Sounds like it will break restore, which might be just fine - but we probably should document it somewhere..

**Comment by Snorch** [comment]
@mclapinski, I wrote a test to show you the problem, see here https://github.com/Snorch/criu/tree/child_subreaper_and_reparent

Before your "Add support for migrating CHILD_SUBREAPER prctl" patch there is no fail in these test, after the patch though we have:
```
[root@snorch criu]# ./test/zdtm.py run -f uns -t zdtm/static/child_subreaper_and_reparent
=== Run 1/1 ================ zdtm/static/child_subreaper_and_reparent
============= Run zdtm/static/child_subreaper_and_reparent in uns ==============
Start test
./child_subreaper_and_reparent --pidfile=child_subreaper_and_reparent.pid --outfile=child_subreaper_and_reparent.out
Run criu dump
Run criu restore
Send the 15 signal to  73
Wait for zdtm/static/child_subreaper_and_reparent(73) to die for 0.100000
###### Test zdtm/static/child_subreaper_and_reparent FAIL at result check ######
Test output: ================================
11:33:15.590:     4: FAIL: child_subreaper_and_reparent.c:67: Parent missmatch before 1 after 5 (errno = 10 (No child processes))

 <<< ================================
##################################### FAIL #####################################
```
That is because you restore CHILD_SUBREAPER _before_ tasks are reparented and they get wrong parent. (see prepare_pstree_ids for reference)

 >Restoring child subreaper attribute after forking will work only on kernels containing this fix so starting from v4.11. For older kernels restoring this attribute after forking will have no effect on children this process already has.

I think we should not restore CHILD_SUBREAPER on kernels before fix at all, because we can't diferentiate children before the flag set and children after it. These way we are free to set the flag as late as possible to not affect process tree topology restore.

**Comment by mclapinski** [comment]
@Snorch You are right about the problem, thanks for the test and explanation.

> I think we should not restore CHILD_SUBREAPER on kernels before fix at all, because we can't diferentiate children before the flag set and children after it. These way we are free to set the flag as late as possible to not affect process tree topology restore.

Ok, I can do that but in Travis we use Xenial, which I think means Ubuntu 16.04.0 therefore v4.4 kernel. If that's the case my test will fail since v4.4 doesn't have your patch to kernel. What do you think?

**Comment by mclapinski** [comment]
> Ok, I can do that but in Travis we use Xenial, which I think means Ubuntu 16.04.0 therefore v4.4 kernel. If that's the case my test will fail since v4.4 doesn't have your patch to kernel. What do you think?

Oh, that's actually not true, since in my test I'm just checking if child subreaper attribute is set, not if it was properly set up for children. But if someone was to write such a test it would fail. Should I care?

**Comment by Snorch** [comment]
@mclapinski Looks like travis kernel should have the fix:
[snorch@snorch ~]$ curl https://api.travis-ci.org/v3/job/555886995/log.txt 2>/dev/null | grep "Runtime kernel version"
Runtime kernel version: 4.15.0-1032-gcp

Not sure if our docker template can/do change kernel version shown inside, but I doubt it does.

You can add new kdat check to enable/disable CHILD_SUBREAPER logic. Either kernel version check (which is not 100% accurate) or create three processes SUBREAPER, HELPER and CHILD like I do in test, set the subreaper flag on the SUBREAPER, kill HELPER and check CHILD is or is not a child of SUBREAPER after these.

**Comment by mclapinski** [comment]
> You can add new kdat check to enable/disable CHILD_SUBREAPER logic. Either kernel version check (which is not 100% accurate) or create three processes SUBREAPER, HELPER and CHILD like I do in test, set the subreaper flag on the SUBREAPER, kill HELPER and check CHILD is or is not a child of SUBREAPER after these.

That sounds pretty bad actually. Can't I just restore it after fork also for kernels without the fix? Before my change the behaviour was broken (one would lose child_subreaper property for both existing and new children) and after my change it will also be broken (one would lose child_subreaper property for existing children but keep it for new children). I don't think the situation after my change is worse than before my change.

Btw. I pushed new code that restores this property after forking. Please review it if you can.

File: test/zdtm/static/child_subreaper_existing_child.c [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+	/* Cleanup: reap the SUBREAPER */
+	waitpid(pid, &status, 0);
+	if (!WIFEXITED(status) || WEXITSTATUS(status)) {
+		fail("Wrong exit status: %d", status);
+		return 1;
+	}
+
+	if (pid != sh->parent_after_cr)
+		fail("Parent missmatch before %d after %d", pid, sh->parent_after_cr);
```
Comment (Snorch) [inline test/zdtm/static/child_subreaper_existing_child.c]: Confusing message here, you check if parent after orphan re-parent is expected (is a SUBREAPER). And maybe change naming for parent_after_cr too.

**Comment by Snorch** [comment]
Except a small nit in test these looks good to me, after fixing you can add

Reviewed-by: Pavel Tikhomirov <ptikhomirov@virtuozzo.com>

Also you can add child_subreaper_and_reparent test to your series if you like (with similar rework you did in child_subreaper_existing_child to make it look better).

note: CR_STATE_RESTORE_SIGCHLD looks a proper place for a flag restore when all helpers and zombies already dead and all reparenting finished.

**Comment by mclapinski** [comment]
@Snorch done

File: criu/pie/restorer.c [hunk 2]
```diff
@@ -1711,6 +1723,10 @@ long __export_restore_task(struct task_restore_args *args)
 	if (ret)
 		goto core_restore_end;
 
+	ret = restore_child_subreaper(&args->child_subreaper);
```
Comment (Snorch) [inline criu/pie/restorer.c]: Sorry, my bad, I confused you with my previous message... what I meant in my note that the place you called restore_child_subreaper was absolutely right. Should've wrote ``` _after_ CR_STATE_RESTORE_SIGCHLD looks a proper place```

File: test/zdtm/static/child_subreaper.c [hunk 3]
```diff
@@ -0,0 +1,36 @@
+#include <stdlib.h>
```
Comment (Snorch) [inline test/zdtm/static/child_subreaper.c]: And it looks like you accidentally merged zdtm test patch with criu patch. Please make them separate patches.

File: test/zdtm/static/child_subreaper_and_reparent.c [hunk 4]
```diff
@@ -19,132 +19,120 @@ enum {
...
+	futex_wait_until(&sh->fstate, TEST_EXIT);
 	return 0;
 }
 
-int main(int argc, char **argv)
-{
-	void *ptr;
-
-	BUG_ON(sizeof(*fstate) + sizeof(*parent_before_cr) + sizeof(*parent_after_cr) > 4096);
+int main(int argc, char **argv) {
```
Comment (Snorch) [inline test/zdtm/static/child_subreaper_and_reparent.c]: We follow linux coding style. Please place beginning brace at the new line for functions.
https://www.kernel.org/doc/html/v4.10/process/coding-style.html#placing-braces-and-spaces

**Comment by Snorch** [comment]
And for child_subreaper_and_reparent, there is no need in two patches, merge your changes with the initial patch.

**Comment by mclapinski** [comment]
@Snorch done

**Comment by Snorch** [comment]
Looks good to me.

File: criu/include/restorer.h [hunk 5]
```diff
@@ -50,6 +50,11 @@ struct restore_posix_timer {
 	int overrun;
 };
 
+struct restore_child_subreaper {
+	bool has_child_subreaper;
```
Comment (avagin) [inline criu/include/restorer.h]: why do you need this has_child_subreaper? child_subreaper is PID, so it is always greater than 0. If child_subreaper is zero, child_subreaper isn't set. Hm?

Comment (mclapinski) [reply criu/include/restorer.h]: `child_subreaper` is not PID.

Comment (avagin) [reply criu/include/restorer.h]: Ok. It is even better:). child_subreaper can be 0 or 1. If it is 0 we need to do nothing. If it is one we need to call prctl(PR_SET_CHILD_SUBREAPER). I still don't understand why we need has_child_subreaper.

**Comment by avagin** [comment]
Applied. Thanks!

**Comment by mclapinski** [comment]
Thanks!
